Refactor ECS to use the Sparse Set data structure

AlecTroemel opened the pull request:
Closes #16 

the only incompatibility here is that I **removed** the ability for components to have single fields. So `(def-componet pizza :string)` is no more. The reason is that its impossible to update those fields, since they're often simple data types which are passed by value! This is actually broken in the `01-ecs.janet` example with the lives component.  

#### TODO:
- [X] get sparse set working on `examples/01-ecs.janet`  
- [x] cache view results. need to invalidate relevant cache whenever a component or entity is added/removed
- [x] either remove component limit, or find clean api to configure it. Should we try to reuse stale entity id numbers??
- [x] verify all other examples and code works 
- [x] create a minimal performance comparison

AlecTroemel committed:
tiny implementation of sparse set on integers

AlecTroemel committed:
changes to sparse set to handle components, implimentation of views

AlecTroemel committed:
integrate sparse set into ecs!

AlecTroemel committed:
remove testing file

AlecTroemel committed:
use identity fn and dont run system if all queries are empty

AlecTroemel committed:
move sparse set into ecs file to work around my non understanding of macros

AlecTroemel committed:
bunch of bug fixes

AlecTroemel committed:
remove dead code

AlecTroemel committed:
tweak comments

pepe commented:
I am on the phone ATM and it looks good, but I can do a proper review tomorrow first thing in the office if you want to. Let me know, please

AlecTroemel commented:
@pepe you're input is always appreciated! However I will say this MR is not quiet ready yet. Specifically, adding caching I suspect will change the code here up pretty significantly. I also am probably going to bite the bullet and write some tests, which could make reviewing easier?

All that to say, maybe hold off until I work through more of the checklist?

pepe commented:
Oh. Thanks for the info! Maybe consider changing PR status to draft? May I ask, is the checklist visible somewhere?

AlecTroemel commented:
hmm, the todo should be right in the description of this PR. I did edit the desc after creating the PR. I also just updated the title to include "DRAFT"

AlecTroemel committed:
split out sparse set into its own file again, fix imports

AlecTroemel committed:
add simple cache of views

AlecTroemel commented:
added a super simple caching of views results. Some super simple timing looks _very promising_ :smile: 

```clojure
(use /junk-drawer)

(def-component position :x :number :y :number)
(def-component velocity :x :number :y :number)

(def world (create-world))

(for i 0 10000
  (add-entity world
              (position :x i :y i)
              (velocity :x i :y i)))

(def-system a {es [:position]} nil)
(def-system b {es [:velocity]} nil)
(def-system c {es [:position :velocity]} nil)

(register-system world a)
(register-system world b)
(register-system world c)

(for i 0 1000
  (:update world 1))
```

timings 
```bash
# main branch
> time janet perf.janet
janet perf.janet  47.26s user 0.01s system 99% cpu 47.276 total

# Sparse set
> time janet perf.janet
janet perf.janet  0.06s user 0.00s system 99% cpu 0.063 total
```

However I still need to invalidate the cache on entity/component adding/removing. I doubt that will drastically change timingresults though!

AlecTroemel committed:
invalidate caches when relevant components change

pepe commented:
It's a pity I blame the phone GH app 😂. Great work anyway, more from the table with display, a real keyboard, coffee, and glasses. So not to embarrass me even more 😉

pepe commented on junk-drawer/cache.janet:
```diff
@@ -0,0 +1,18 @@
+(defn init []
+  (table/setproto
+   @{}
+   @{:insert cache-insert
```
I would remove the `cache-` prefix from the function names.

AlecTroemel replied:
Yaaa, I coded myself into a corner by naming one of the functions `get`, and I didn't want to override the standard API. The functions are small enough that I could probably get away with in-lining them right into the prototype object, but that kind of feels wrong also.

pepe replied:
That makes sense. If functions will not be part of the public API I am using good old _ at the start of the name in these cases, together with `defn-`. Which indeed does not feel great too

AlecTroemel replied:
ah ya totally forgot to use `defn-`, `_get` is good to me!

pepe commented on junk-drawer/ecs.janet:
```diff
@@ -1,92 +1,135 @@
+(import /junk-drawer/sparse-set :as sparse-set)
```
The `:as` is redundant here.

pepe commented on junk-drawer/ecs.janet:
```diff
@@ -1,92 +1,135 @@
+       (when (nil? (get-in ,$wld [:database ,$cmp-name]))
+         (put-in ,$wld
+                  [:database ,$cmp-name]
+                  (sparse-set/init MAX_ENTITY_COUNT MAX_ENTITY_COUNT)))
```
I would unquote the `sparse-set/init` function here, which could solve some of the code shuffling in this PR. It also applies to some more functions in the quasi quotes down the road.

AlecTroemel replied:
I had a lot of problems with this macro, so I'll definitely try that out!

pepe replied:
It is one of the very good Janet  macro tricks

pepe reviewed (commented):
Please take this as my point of view. The thing is, most of the recommendations are more of my way of doing things in Janet. Also, I do not have any Janet projects to review code in ATM :-D. I hope you will take this friendly with an open mind.

There are two more,  which are repeated in your code:

- I always try to format docstring as prose, so sentences start with capital letters and end with the period.
- I am always using `use` where I do not care about the prefix of the import. If for no other reason but the variable arity of the macro https://github.com/janet-lang/janet/blob/8f0a1ff/src/boot/boot.janet#L2948.

AlecTroemel committed:
remove redundant :as

AlecTroemel committed:
mark cache functions as private, remove cache in name and just use _ as needed

AlecTroemel committed:
unquote sparse-set/init in add-component macro

no longer need to expose the sparse set in junk-drawer.janet

AlecTroemel committed:
no longer need to import sparse set in timers

AlecTroemel committed:
reuse id's that have been removed

pepe commented on junk-drawer/ecs.janet:
```diff
@@ -1,92 +1,135 @@
+       (when (nil? (get-in ,$wld [:database ,$cmp-name]))
+         (put-in ,$wld
+                  [:database ,$cmp-name]
+                  ,(sparse-set/init MAX_ENTITY_COUNT MAX_ENTITY_COUNT)))
```
I am sorry, but I meant to unquote the function, not the result of the call

AlecTroemel replied:
you mean this? :exploding_head:

```clojure
(,sparse-set/init MAX_ENTITY_COUNT MAX_ENTITY_COUNT)
```

AlecTroemel committed:
split out max entity count from id, actually quote the ss init fn

AlecTroemel committed:
remove max entity count limitation, allow passing capacity through world init

AlecTroemel committed:
fix function call in example

AlecTroemel committed:
put the view function in the world object, use it to fix example 6

AlecTroemel committed:
add docs strings in proper english to all ECS related functions

AlecTroemel committed:
add a complex stress test that touches most of the ECS

AlecTroemel commented:
I've made a couple of updates here

- made sure all the docstrings start with capital letters and end with the period. I also reworded some things to hopefully be more clear.
- `use` is used where it can be, thanks to the `quote the function` deep magic
- added a `test/performance/stress-test.janet` script that has horrible macros but tests more parts of the ECS. Seems like there's a bottleneck creating/deleting entities from the world. However the difference in perf between this and main is still about the same

```bash
# main
janet test/performance/stress-test.janet  171.47s user 0.02s system 99% cpu 2:51.52 total

# sparse set
janet test/performance/stress-test.janet  2.90s user 0.02s system 99% cpu 2.923 total
```

pepe commented on test/performance/stress-test.janet:
```diff
@@ -0,0 +1,76 @@
+
+(use /junk-drawer)
+
+(var ALPHABET ["a" "b" "c" "d" "e" "f" "g" "h" "i"
```
I guess this could be `def`ed as a constant?

AlecTroemel replied:
resolving these comments, since you fix 'em https://github.com/AlecTroemel/junk-drawer/pull/22

pepe commented on test/performance/stress-test.janet:
```diff
@@ -0,0 +1,76 @@
+
+(defmacro def-systems-alphabet []
+  "Define and register a system with 2 queries, each with 2 alphabet components."
+  ~(upscope
```
I have not tried it, I hope I will have time to try it later today, but I think `upscope`s are not needed here.

pepe commented on test/performance/stress-test.janet:
```diff
@@ -0,0 +1,76 @@
+  ~(upscope
+    # define system
+    ,;(map (fn [i]
+             (tuple ~def-system (symbol "sys-" i)
```
Single symbols are better `quote`d with `'`. There is no need to `quasiquote` them with `~`.

pepe commented on junk-drawer/timers.janet:
```diff
@@ -1,4 +1,4 @@
-(use ./ecs)
```
This change goes against the stated move to `use`. May I ask what the reasons are? I am asking out of curiosity.

AlecTroemel replied:
huh, I  thought I changed these back to `use` :facepalm: , must have forgot to commit it. I'll blame git

AlecTroemel replied:
imports back to `use`

pepe reviewed (commented):
I left some minor comments, yet as a whole, I am delighted with these many improvements to the core functionality. I will try to check the code running on my machine today. Thank you!

pepe committed:
refactor stress test macros

Some code design changes:
- remove upscopes
- do not quote result arrays
- change quasiquote ~ to quote ', where it is not needed
- make ALPHABET constant
- change log prints from pp to print and add final one

pepe committed:
Refactor lambdas

pepe committed:
One more inc

pepe committed:
Remove false macro claim :-D

AlecTroemel committed:
Merge pull request #22 from pepe/sparse-sets

refactor stress test macros

AlecTroemel committed:
switch import back to "use"

pepe commented:
Thank you very much for considering my codes!

The one thing I have been wondering about from the beginning of this adventure: What if we make sparse-set Janet abstract type in C. After this PR gets in and settled, would you consider checking that route?

AlecTroemel commented:
yes, I've actually been thinking about that a lot also! 

One concern would be loosing any sort of portability. Consider if someone eventually added Janet to [TIC-80](https://github.com/nesbox/TIC-80/). With a pure Janet implementation you could copy/paste the code here and be on your way. But with C bindings I don't know if/how that would work. However that's a very specific concern for something that doesn't exist yet, and since right now I'm more interested in doing stuff with Jaylib/Raylib its definitely not a deal breaker.

Another thought would be if were writing C code, would it be a _ton more work_ to go the extra mile and bind to an existing ECS like FLECS?.. I'm not sure that even fits the philosophy of this code base, plus my gut says I'm underestimating the work there haha. 

TL;DR: Yes, I'm open to moving the sparse set, and maybe even the view/query logic, to C!

AlecTroemel committed:
add testament library and tests for cache

pepe commented:
That is heavenly chords to my ears. I would even go one step further and write it in `jpm/cgen`.

As for wrapping some other code, I am not interested in this approach as I would like to stay in Janet's realm mentally

AlecTroemel committed:
fix some bugs in delete/clear sparse set fn, add tests for sparse set

AlecTroemel committed:
use janet profiling for stress test

AlecTroemel committed:
switch to spork for both unit tests and performance timing

AlecTroemel committed:
Merge pull request #24 from AlecTroemel/tests

Add unit tests for sparse-set and cache, use spork/timeit for more granular...timing

AlecTroemel commented:
I plan on merging this today or tomorrow!
